main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala

```diff
@@ -652,4 +651,235 @@ object SetFormula {
+    *
+    * Invariant: [[ElemSet]], [[Cst]], and [[Var]] must use disjoint integers.
+    */
+  def propagation(t: SetFormula): SetFormula = propagationHelper(t, SortedMap.empty)
+
+  /**
+    * Applies set propagation to `f` where `insts` keep track of running instantiations for
+    * [[ElemSet]], [[Cst]], and [[Var]] (using the identity mapping if not present).
+    */
+  private def propagationHelper(f: SetFormula, insts: SortedMap[Int, UnivOrEmpty]): SetFormula = f match {
+    case Univ => f
+    case Empty => f
+    case Cst(c) => insts.getOrElse(c, f)
+    case Var(x) => insts.getOrElse(x, f)
+    case e@ElemSet(_) => instElemSet(e, insts)
+    case compl@Compl(f1) =>
+      val f1Prop = propagationHelper(f1, insts)
+      // Maintain and exploit reference equality for performance.
+      if (f1Prop eq f1) compl else mkCompl(f1Prop)
+
+    case Inter(elemPos0, cstsPos0, varsPos0, elemNeg0, cstsNeg0, varsNeg0, rest0) =>
+      // Compute element sets and check for early exits.
+      val elemPos = elemPos0.map(instElemSet(_, insts)) match {
+        case Some(e) => e match {
+          case Univ => None
+          case Empty => return Empty
+          case e@ElemSet(_) => Some(e)
+          case _ => ??? // Unreachable according to instElemSet.
+        }
+        case None => None
+      }
+      if (cstsPos0.exists(c => insts.get(c.c).contains(Empty))) return Empty
+      if (varsPos0.exists(x => insts.get(x.x).contains(Empty))) return Empty
+      val elemNeg = elemNeg0.map(instElemSet(_, insts)) match {
+        case Some(e) => e match {
+          case Univ => return Empty
+          case Empty => None
+          case e@ElemSet(_) => Some(e)
+          case _ => ??? // Unreachable according to instElemSet.
+        }
+        case None => None
+      }
+      if (cstsNeg0.exists(c => insts.get(c.c).contains(Univ))) return Empty
+      if (varsNeg0.exists(x => insts.get(x.x).contains(Univ))) return Empty
+
+      // Compute constants and variables by removing constants and variables that are redundant.
+      // We know from previous checks that none are mapped to the respective short-circuit element.
+      val cstsPos = cstsPos0.filterNot(c => insts.get(c.c).contains(Univ))
+      val varsPos = varsPos0.filterNot(x => insts.get(x.x).contains(Univ))
+      val cstsNeg = cstsNeg0.filterNot(c => insts.get(c.c).contains(Empty))
+      val varsNeg = varsNeg0.filterNot(x => insts.get(x.x).contains(Empty))
+
+      // Add new instantiations.
+      var currentInsts = insts ++
+        setElemOneOpt(elemPos, Univ) ++
+        setElemOneOpt(elemNeg, Empty) ++
+        cstsPos.iterator.map(_.c -> Univ) ++
+        cstsNeg.iterator.map(_.c -> Empty) ++
+        varsPos.iterator.map(_.x -> Univ) ++
+        varsNeg.iterator.map(_.x -> Empty)
+
+      // If a subformula becomes an element set, a constant, or a variable after propagation,
+      // add it to the propagated atoms. This information could also be propagated backwards for
```

Comment:
Could be?  Is this a TODO or something the code does? It leaves me confused.

Reply:
ill remove the could be part